### DISCUSSION TITLE
How to upload an entire folder?

### DISCUSSION THREAD

**Discussion by Shinehah-Gnolaum** [community/community#21962] [New to GitHub]
<p>I clicked on my repo in a browser, then I clicked on Upload Files. I would like to upload an entire folder so that GitHub would preserve the hierarchy of the files and folders in my folder, but I am not able to select folders to be uploaded. How would I upload an entire folder of an XCode project?</p>

**Comment by lee-dohm** [comment]
<p>There isn’t a way to upload an entire folder through the web UI. The easiest way to do that would be to:</p>
<ol>
<li>Clone the repository locally</li>
<li>Make the changes to the local version</li>
<li>Commit the changes locally</li>
<li>Push the changes back up to the GitHub repository</li>
</ol>
<p>You can find out <a href="https://help.github.com/articles/adding-a-file-to-a-repository-using-the-command-line/" rel="nofollow noopener">more information about this process</a> in our help documentation. Let us know if you have any questions.</p>

**Reply by uchiha-no-sharingan** [reply]
That's right. You have to use Github Desktop to do it. The instructions are in this thread :
https://stackoverflow.com/questions/40720141/how-to-upload-folders-on-github

Using the CLI is also possible, but you might run into errors.

**Reply by mkos11** [reply]
Absolutely! Thanks for the guidance. I'll follow these steps to upload the folder. If any questions pop up, I'll be sure to reach out. Appreciate the assistance! 👍✨

**Reply by nassahkj** [reply]
try using the following
cd namedlocalrep
git init
git add :/ <this stages all the files and folders within that that active directory>
git commit -m    "Your msg" -a 
git remote add origin https://github.com/yourUsername/yourRepository.git
git push origin master

**Reply by faulknor82** [reply]
Unfortunately, GitHub refuses to allow me to commit anything because line endings don't match.

**Reply by photobangkok** [reply]
how about shift-select the whole bunch of files?

**Reply by Anirban-Manna-10** [reply]
Try to copy the repository then upload the folder on ur local , then add , commit , push .. I dont know how but it worked for me .

**Reply by hammadshub** [reply]
go in your respected repo , click on Add file button then click on upload files option.... then drag your folder in  UI..... that worked for me , hopefully it will work for everyone!

**Reply by srimathi-dev** [reply]
> go in your respected repo , click on Add file button then click on upload files option.... then drag your folder in UI..... that worked for me , hopefully it will work for everyone!

This is working! thanks!

**Reply by jana602** [reply]
> go in your respected repo , click on Add file button then click on upload files option.... then drag your folder in UI..... that worked for me , hopefully it will work for everyone!

Yes!!

**Reply by ok626** [reply]
yeah u need to drag a folder into web browser, not click on open files and selecting a folder. but seriously guys, stop vibecoding nonsense

**Comment by ra-hulgupta** [comment, marked as answer]
<p>you should just try to drag and drop your entire folder in the webui and it will upload whole folder.</p>

**Reply by foopledoop** [reply]
this no work :(

**Reply by Haresh2122** [reply]
Tried this but isn't working.

**Reply by 0kh0** [reply]
It doesn't work. =(

**Reply by rhysisthebestboi** [reply]
it no workies

**Reply by taharshd** [reply]
it is worky the denier just do it badder

**Reply by aniketmlk6** [reply]
It worked for me!
Thanks!

**Reply by EJD799** [reply]
doesn't work

**Reply by RamSankarTheDeveloper** [reply]
works for me (I'm using web version)

**Reply by Shaffer02** [reply]
cool, it worked.

**Reply by hh8385** [reply]
thanks, it worked for me

**Reply by m4heshd** [reply]
This works absolutely fine. Why are people so tilted about this solution? 🤷🏽

**Reply by Arpit65-sys** [reply]
but this upload all files of folder one by one

**Reply by captainwycliffe** [reply]
Thanks it worked for me!

**Reply by smiller7812** [reply]
Doesn't work for me.

**Reply by booger23** [reply]
I had the same issue while using Chrome. I switched to Brave and it worked fine

**Reply by farhanrejwan** [reply]
This "drag the folder and drop it to the web UI" doesn't let you upload the folder if there are more than 100 files in it.
It says, "Yowza, that’s a lot of files. Try uploading fewer than 100 at a time.".

**Reply by erikiera** [reply]
This doesn't work.

**Reply by hasib101101** [reply]
You should have explained a bit more

**Reply by mahmoudsharifi** [reply]
It worked for me. I have uploaded folders one by one through UI and it is working. Thanks.

**Reply by ghost** [reply]
Wow, saved me hours, was adding folders individually

**Reply by mkos11** [reply]
Absolutely! Thanks for the guidance. I'll follow these steps to upload the folder. If any questions pop up, I'll be sure to reach out. Appreciate the assistance! 👍✨

**Reply by ali-asad** [reply]
what if i have a parent folder and the child folders and child folders contains more than 100 files in total. How can i intact the hierarchy of files in my project ?

**Reply by ghost** [reply]
Just drag and drop the parent and the children will come too.

**Reply by RBRK6270** [reply]
For me too. Even ZIP folder is working. When more than 100 files it is not allowing if we ZIP it all can be uploaded

**Reply by CompileMyLife** [reply]
Thank you that worked perfectly!

**Reply by pers-ephone** [reply]
- The HTML `<input type="file">` element does not allow the selection of directories.  
- The drag and drop works with js and will accept directories under 100 files total.  
- For more you need to clone the repo or [use `github.dev`](https://github.com/orgs/community/discussions/21962#discussioncomment-9211215)

(just putting this recap out there)

**Reply by m4heshd** [reply]
### For everyone who's absolutely tripping over this issue, simply follow these steps: 

1. Open up the repo
2. Edit the domain on the address bar from `github.com` to `github.dev` (`github.dev/username/repo`). This will open the VS Code (Monaco) web code editor provided by GitHub, pointing to that repo
3. You can drag and drop anything to the file tree on that editor and commit the changes

Done. 🤷🏻‍♂️

**Reply by Nintendoboi22** [reply]
press "." to open it up faster on any repo

**Reply by m4heshd** [reply]
> press "." to open it up faster on any repo

I didn't mention this because it doesn't work on all browsers and platforms.

**Reply by MTalhapk** [reply]
> Thank you that worked perfectly!

simply zip the entire folder and upload it

**Reply by mikaere66** [reply]
> press "." to open it up faster on any repo

Worked perfectly in Edge for Linux ... thank you =)

**Reply by VaishnaviTyagi23** [reply]
Thanks !!

**Reply by radekszczygiel** [reply]
It works ✅. Thank you!

**Reply by AshutoshPundir** [reply]
yes it worked

**Reply by HMTN357** [reply]
It says I can upload files only or is didn't let me do that.

**Reply by ianiketkumar7** [reply]
Thank you for this. It worked totally fine.

**Reply by Samiruddink** [reply]
but it shows upload fewer files

**Reply by Space-coww** [reply]
Didn't work atleast for firefox

**Reply by chris4sun01** [reply]
it didn't work

**Reply by Alethiea-Cheng** [reply]
it didn't work on 5/14 for chrome

**Reply by Alethiea-Cheng** [reply]
@chris4sun01  hey it still works for microsoft edge. Use edge browser

**Reply by dejvikocilja** [reply]
it worked, but not for all files/folders within the main folder

**Reply by yaswanthjemudugani** [reply]
It's working for me in Chrome browser.

<img width="1292" height="691" alt="image" src="https://github.com/user-attachments/assets/2946f21f-9049-4d25-afbd-591d7eb85969" />

<img width="1292" height="691" alt="image" src="https://github.com/user-attachments/assets/98cd2d08-ee32-4334-8423-69d6ec898970" />


Thankyou @ra-hulgupta

**Reply by The-Final-Apex** [reply]
Oh no... was it that easy this whole time?!?

**Comment by nadiajoyce** [comment]
<p>Wanted to confirm here, that dragging and dropping a folder into the Web UI will now, indeed, create that folder in (and upload the files within it to) your GitHub repo. See: <a href="https://help.github.com/articles/adding-a-file-to-a-repository/" rel="nofollow noopener">https://help.github.com/articles/adding-a-file-to-a-repository/</a> for more info.</p>

**Reply by smiller7812** [reply]
This is not working for me.

**Reply by farhanrejwan** [reply]
This "drag the folder and drop it to the web UI" doesn't let you upload the folder if there are more than 100 files in it.
It says, "Yowza, that’s a lot of files. Try uploading fewer than 100 at a time.".

**Reply by MattSayar** [reply]
I tried uploading a directory called ".github" and it says "This file is hidden" and won't continue

**Reply by EvilDagger** [reply]
Try making it visible maybe? Change the filename from .github to github. It will certainly be annoying for a reason or two but it should work.

**Comment by Siva-TN** [comment]
<p>I have tried to upload a folder with drag and drop option but it’s not working. So can you please suggest me how to upload an entire folder?</p>

**Reply by mkos11** [reply]
Absolutely! Thanks for the guidance. I'll follow these steps to upload the folder. If any questions pop up, I'll be sure to reach out. Appreciate the assistance! 👍✨

**Reply by STPv22** [reply]
Some here

**Reply by Assam157** [reply]
Just make a zip and upload

**Reply by hammadshub** [reply]
> I have tried to upload a folder with drag and drop option but it’s not working. So can you please suggest me how to upload an entire folder?

go in your respected repo , click on Add file button then click on upload files option.... then drag your folder in  UI..... that worked for me , hopefully it will work for everyone!

**Reply by Space-coww** [reply]
> Just make a zip and upload

are we fr

**Reply by NupoorsCode** [reply]
> > I have tried to upload a folder with drag and drop option but it’s not working. So can you please suggest me how to upload an entire folder?
> 
> go in your respected repo , click on Add file button then click on upload files option.... then drag your folder in UI..... that worked for me , hopefully it will work for everyone!

thanks, this worked

**Comment by nadiajoyce** [comment]
<p><a class="mention" href="/siva-tn">@siva-tn</a> Is the repo you’re trying to upload the folder to public? If so, can you share the link to it? Also, what file formats are contained in the folder you’re trying to upload? </p>

**Reply by mkos11** [reply]
Absolutely! Thanks for the guidance. I'll follow these steps to upload the folder. If any questions pop up, I'll be sure to reach out. Appreciate the assistance! 👍✨

**Comment by MateLaszloToth** [comment]
<p>Or use command line:</p>
<p>cd into the directory where your folder is located</p>
<p>type:</p>
<pre><code>git init
git add &lt;folder1&gt; &lt;folder2&gt; &lt;etc.&gt;
git commit -m "Your message about the commit"
git remote add origin https://github.com/yourUsername/yourRepository.git
git push -u origin master
git push origin master
</code></pre>

**Reply by Lerazor34** [reply]
thanks g

**Reply by B21q** [reply]
When I get to that part, it asks for my credentials, even putting it right says it doesn't work.

**Reply by m4heshd** [reply]
@B21q Do you have 2FA enabled for your GitHub account? If that's the case, you must use a Personal Access Token.

**Reply by B21q** [reply]
I don't have

**Reply by m4heshd** [reply]
Still, I think GitHub is enforcing it now. You NEED to use a PAT. They won't allow password logins now.

**Reply by B21q** [reply]
Thanks :)

**Reply by smiller7812** [reply]
After I do git init, 
I type git add Module16.
It says, "error: 'Module16/FredsCars does not have a commit checked out.
fatal: adding files failed."

**Reply by wealthczarena** [reply]
Try git add -A

**Reply by AjitChauhan081** [reply]
try deleting .gitignore

**Reply by mkos11** [reply]
Absolutely! Thanks for the guidance. I'll follow these steps to upload the folder. If any questions pop up, I'll be sure to reach out. Appreciate the assistance! 👍✨

**Reply by infinitestrider** [reply]
pushing from terminal does not work for me iam on linux and all it says is authentication no longer uses pass word auth, iam trying to make a cms with a bunch of open sourch cms non of them worked because of the git push bs. now i tried uploading uploading by drag and dropping and it says yowza thats alot of files, what do they want us to do lol

**Reply by infinitestrider** [reply]
alright i managed to get it done, by searching my issue on youtube.

**Reply by nnewell90** [reply]
@infinitestrider what youtube video did you find to resolve this issue? I am currently having the same problem and cannot find a way around the "no longer uses password auth". I've tried gpg key as well and I still cannot get around it. Any help is appreciated!

**Reply by infinitestrider** [reply]
When making a PAT make sure its the classic one, and make sure you give it all permissions so click everything and check everything to make sure it has support for everything then use that Personal Access Token, copy and paste it after you give your username. The permissions thing is important apparently.

**Comment by ghost** [comment]
<p>I have tried to drag the folder but it shows folder contains more than 100 files, try with less number of files, is there another way to do so?, i have nearly train_data images in my folder.</p>

**Reply by Git-happiness** [reply]
This was the case for me too. I uploaded each folder separately .

**Comment by veskoaleksandrov** [comment]
<p>Didn’t work  in FF. GC worked fine though…</p>

**Reply by mkos11** [reply]
Absolutely! Thanks for the guidance. I'll follow these steps to upload the folder. If any questions pop up, I'll be sure to reach out. Appreciate the assistance! 👍✨

**Comment by advaymehta009** [comment]
<p>not working, only files were uploading not entire folder.</p>

**Reply by mkos11** [reply]
Absolutely! Thanks for the guidance. I'll follow these steps to upload the folder. If any questions pop up, I'll be sure to reach out. Appreciate the assistance! 👍✨

**Comment by zohaibaamer766** [comment]
<p>how to set path format in git add &lt;folder&gt;</p>

**Comment by TanmaySM** [comment]
<p>Folder upload only works if contains less than 100 files.</p>

**Reply by Mohammad-Rifat-Khan** [reply]
what should i do if it has over 100 files?

**Reply by farhanrejwan** [reply]
upload the files individually bruh :/

**Comment by joelzehring** [comment]
<p>I’m trying to drag and drop the build folder of a small react app and it doesn’t let me drop it in the web interface. Seems to be less than 100 files in the directory… Any idea what the problem might be?</p>

**Reply by mkos11** [reply]
Absolutely! Thanks for the guidance. I'll follow these steps to upload the folder. If any questions pop up, I'll be sure to reach out. Appreciate the assistance! 👍✨

**Comment by ptoye** [comment]
<p>Same here (once I’d worked out whatr FF and GC were). Something odd in Firefox I guess.</p>

**Comment by loganchen39** [comment]
<p>I have this problem too, I’ve created a new Repo on my Windows 10 Github Desktop, and it won’t upload some directories (e.g. build, lib) to <a href="http://github.com" rel="nofollow noopener">github.com</a>, the directory lib only has one binary file, can someone help how to solve this problem? Thanks,</p>

**Comment by gitPrabhuDutt** [comment]
<p>Thanks man, It worked.</p>

**Comment by Jean-Laurore** [comment]
<p>My Error:</p>
<p>remote: Permission to Jean-Laurore/BookLibrary.git denied to jelaurore.</p>
<p>fatal: unable to access ‘<a href="https://github.com/Jean-Laurore/BookLibrary.git/" rel="nofollow noopener">https://github.com/Jean-Laurore/BookLibrary.git/</a>’: The requested URL returned error: 403</p>

**Comment by sahumanika** [comment]
<p>Using the ‘Upload files’ button allows uploading fewer than 100 files.</p>
<p>Try GIT Bash, use the following codes to upload the whole folder.</p>
<ol>
<li>Open Git Bash.</li>
<li>CD projectname</li>
<li>$ git init</li>
<li>$ git add .</li>
<li>$ git commit -m “First commit”</li>
<li>$ git remote add origin remote repository URL</li>
<li>$ git remote -v [Note: Verifies the new remote URL]</li>
<li>$ git push origin master</li>
</ol>
<p>Thank you!</p>
<p>Manika Sahu</p>

**Reply by muzakkimz89** [reply]
thanks, it work

**Reply by wealthczarena** [reply]
thanks for sharing, it works!

**Comment by SanyaAhmed7** [comment]
<p>Hey there, how can I set repository in cmd when my git folder lies on desktop?</p>

**Comment by gurjinderbassi** [comment]
<p>But it doesn’t allow to upload files larger than 25 Mb</p>

**Comment by jfullstackdev** [comment]
try GitHub Desktop

still the same rules apply though in file limit sizes

**Reply by bharaniyhao** [reply]
yeah, Github should allow upload of folders

**Comment by YoloNet** [comment]
You can only upload folder if it contains less than 100 Files. Drag & Drop

**Comment by AbeerRoy** [comment]
Open Git Bash.
cd _projectdirectory_
1. $ git init
2. $ git add _foldername_
3. $ git commit -m _“message”_
4. $ git config --global user.email "_you@example.com_"
or 
4. $ git config --global user.name "_Your Name_"
5. $ git push

Authenticate with your credentials just by SignIn in the git pop-up

**Comment by andrewmuhoro** [comment]
By drag and dropping, it worked for me. Ensure you have a repository already created on Github though.

**Reply by bharaniyhao** [reply]
you sure? have try it but only works on files

**Reply by Git-happiness** [reply]
> you sure? have try it but only works on files

Make sure the folder/folders contain less than 100 files. There will be a warning if they do.

**Reply by bharaniyhao** [reply]
okay

**Reply by Fitsneezy** [reply]
> > you sure? have try it but only works on files
> 
> Make sure the folder/folders contain less than 100 files. There will be a warning if they do.

what if it's a unity game?

**Comment by KasperCastelijn** [comment]
when dragging a folder, it does list all the files individually, but after commit it is correctly in the folder(s) as wanted

just do a lil test with a new folder with 2 txt files and see what happens

**Comment by TAIHAN86** [comment]
Doesn't work for me.

**Reply by faiyankhan26** [reply]
hm

**Reply by B21q** [reply]
You need to login using github token

**Comment by nassahkj** [comment]
Alright then

**Comment by m4heshd** [comment]
### There's a modern solution for this. Simply follow these steps: 

1. Open up the repo
2. Edit the domain on the address bar from `github.com` to `github.dev` (`github.dev/username/repo`). This will open the VS Code (Monaco) web code editor provided by GitHub, pointing to that repo
3. You can drag and drop anything to the file tree on that editor and commit the changes

Done. 🤷🏻‍♂️

**Reply by pers-ephone** [reply]
@m4heshd good to know as one more option but i guess the point is simply that nothing states the fact the 'choose your files' link is more limited than any other options

**Reply by m4heshd** [reply]
@pers-ephone Agreed. When GitHub sucks, it sucks hard. GitHub is a great overall platform but this is one of those rare infuriating issues I found on here in my decades of usage. I just completely managed not to run into it by using `github.dev`.

**Comment by nassahkj** [comment]
Thanks for the tip!

**Comment by astrodingra** [comment]
I used to upload the folder by splitting the files in different folder and at them in GitHub to the same folder one by one

**Comment by lumgwun** [comment]
> <p>Thanks man, It worked.</p>

What did you do, how did you upload the folder

**Reply by mehulmina064** [reply]
1. Init a repository or clone the old one
2. add your folder to that place
3. remove it from .gitignore
4. commit and push -> Boom it is uploaded to git if the size is not that much bigger

**Reply by m4heshd** [reply]
https://github.com/orgs/community/discussions/21962#discussioncomment-9211201

**Comment by the5orcerer** [comment]
Go to your repo in browser then create a file, in file name -> Foldername/file.txt
Upload your files and then delete the file.txt

**Comment by Finite-Code** [comment]
Using the terminal is the best solution if you got files of 100mb...

**Comment by pradyumin-coder** [comment]
Compress the folder as .zip file and upload and it will upload it.

**Comment by Perfect-Biz-Map** [comment]
> I clicked on my repo in a browser, then I clicked on Upload Files. I would like to upload an entire folder so that GitHub would preserve the hierarchy of the files and folders in my folder, but I am not able to select folders to be uploaded. How would I upload an entire folder of an XCode project?

It will only work if you drag and drop only one folder at a time from the local PC, no matter how many files or folders are inside that folder. It will not work if you drag and drop two folders at once from the local PC.

**Comment by shawshank1307** [comment]
You can upload an entire folder to GitHub in a few ways:

### Option 1: GitHub website (drag & drop)
1. Open your repository on GitHub
2. Click **Add file → Upload files**
3. Drag the **whole folder** into the upload area (or click to browse)
4. GitHub will upload all files inside it, keeping the folder structure
5. Scroll down, add a commit message, and click **Commit changes**

### Option 2: Git command line (recommended for large folders)
```bash
git clone https://github.com/YOUR_USERNAME/YOUR_REPO.git
cd YOUR_REPO
cp -r /path/to/your/folder ./
git add .
git commit -m "Add folder"
git push
```

### Option 3: GitHub Desktop
1. Clone the repo in GitHub Desktop
2. Copy your folder into the repo directory on your computer
3. You will see all new files listed — commit and push

**Tip:** Git does not track empty folders. Add a `.gitkeep` file inside any empty folder you want to preserve.

Hope this helps!
